OpenRA.Mods.Common/UtilityCommands/ExtractEmmyLuaAPI.cs

```diff
@@ -0,0 +1,358 @@
+					if (methodInfo != null)
+					{
+						var parameters = methodInfo.GetParameters();
+						foreach (var parameter in parameters)
+							Console.WriteLine($"    ---@param {parameter.EmmyLuaString()}");
+
+						var parameterString = parameters.Select(p => p.Name).JoinWith(", ");
+
+						var attributes = methodInfo.GetCustomAttributes(false);
+						foreach (var obsolete in attributes.OfType<ObsoleteAttribute>())
+							Console.WriteLine($"    ---@deprecated {obsolete.Message}");
+
+						var returnType = methodInfo.ReturnType.EmmyLuaString();
+						if (returnType != "Void")
+							Console.WriteLine($"    ---@return {returnType}");
+
+						body = member.Name + $" = function({parameterString}) end;";
+					}
+
+					Console.WriteLine($"    {body}");
+				}
+
+				Console.WriteLine("}");
+				Console.WriteLine();
+			}
+		}
+
+		static void WriteScriptProperties(Type type, IEnumerable<Type> implementingTypes)
+		{
+			var className = type.Name.ToLowerInvariant();
+			var tableName = $"__{type.Name.ToLowerInvariant()}";
+			Console.WriteLine($"---@class {className}");
+			Console.WriteLine("local " + tableName + " = {");
+
+			var properties = implementingTypes.SelectMany(t =>
+			{
+				var required = ScriptMemberWrapper.RequiredTraitNames(t);
+				return ScriptMemberWrapper.WrappableMembers(t).Select(memberInfo => (memberInfo, required));
+			});
+
+			foreach (var property in properties)
+			{
+				Console.WriteLine();
+
+				var isActivity = property.memberInfo.HasAttribute<ScriptActorPropertyActivityAttribute>();
+
+				if (property.memberInfo.HasAttribute<DescAttribute>())
+				{
+					var lines = property.memberInfo.GetCustomAttributes<DescAttribute>(true).First().Lines;
+					foreach (var line in lines)
+						Console.WriteLine($"    --- {line}");
+				}
+
+				if (isActivity)
+					Console.WriteLine("    --- *Queued Activity*");
+
+				if (property.required.Any())
+					Console.WriteLine($"    --- **Requires {(property.required.Length == 1 ? "Trait" : "Traits")}:** {property.required.Select(GetDocumentationUrl).JoinWith(", ")}");
+
+				var methodInfo = property.memberInfo as MethodInfo;
+				if (methodInfo != null)
+				{
+					var attributes = methodInfo.GetCustomAttributes(false);
+					foreach (var obsolete in attributes.OfType<ObsoleteAttribute>())
+						Console.WriteLine($"    ---@deprecated {obsolete.Message}");
+
+					var parameters = methodInfo.GetParameters();
+					foreach (var parameter in parameters)
+						Console.WriteLine($"    ---@param {parameter.EmmyLuaString()}");
+
+					var parameterString = parameters.Select(p => p.Name).JoinWith(", ");
+
+					var returnType = methodInfo.ReturnType.EmmyLuaString();
+					if (returnType != "Void")
+						Console.WriteLine($"    ---@return {returnType}");
+
+					Console.WriteLine($"    {methodInfo.Name} = function({parameterString}) end;");
+				}
+
+				var propertyInfo = property.memberInfo as PropertyInfo;
+				if (propertyInfo != null)
```

Comment:
C# 7 pattern matching is now used as of PR https://github.com/OpenRA/OpenRA/pull/20020:
```suggestion
				if (property.memberInfo is PropertyInfo propertyInfo)
```